// nodechef-cli/endpoints.js
var BARE_METAL_CLOUD_PROD = 'https://nodechef.com/api/',
    BARE_METAL_CLOUD_DEV = 'http://127.0.0.1:8080/api/',
    BARE_METAL_CLOUD_ALIAS = 'nc-bare-metal', 
    VPS_CLOUD_PROD = 'http://vps.nodechef.com/api/',
    VPS_CLOUD_DEV = 'http://localhost:8080/api/', 
    VPS_CLOUD_ALIAS = 'nc-vps', 
    cli = require('./cli'); 

var endpoints = {
    envBM : BARE_METAL_CLOUD_ALIAS,
    appCloudEnv : function(){
        return (process.env.DEVELOP ? BARE_METAL_CLOUD_DEV : BARE_METAL_CLOUD_PROD)
         + 'v1/clouds/environment'; 
    },
    appId: function () {
       return (process.env.DEVELOP ? VPS_CLOUD_DEV : VPS_CLOUD_PROD) + 'instances/app/getappid'; 
    }, 
    baseUrl: function (__env) {
        if (__env == BARE_METAL_CLOUD_ALIAS) {
            return process.env.DEVELOP ? BARE_METAL_CLOUD_DEV : BARE_METAL_CLOUD_PROD; 
        }
        else if (__env == VPS_CLOUD_ALIAS) {
            return process.env.DEVELOP ? VPS_CLOUD_DEV : VPS_CLOUD_PROD;
        }
        else {
            cli.writeerror('The specified environment is not a recognized NodeChef Cloud environment'); 
            process.exit(1); 
        }
    },

    deploy: function (__env) {
        return endpoints.baseUrl(__env) + 
            ((__env == BARE_METAL_CLOUD_ALIAS) ? 'v1/app/deploy' : 'instances/app/deploy');
    },

    bareMetalOnly : function(){
        cli.writeerror('Operation only supported in the NodeChef bare metal cloud environment');
    }, 

    staticsdeploy: function (__env) {
        if (__env != BARE_METAL_CLOUD_ALIAS) {
            endpoints.bareMetalOnly();
            cli.writeerror('To deploy static assets to the VPS Cloud environment, create an app and use the [static] buildpack powered by Nginx'); 
            process.exit(1); 
        }

        return endpoints.baseUrl(__env) + 'v1/app/deploystatics'; 
    },
    cloudcodedeploy: function(__env) {
        if (__env != BARE_METAL_CLOUD_ALIAS) {
            endpoints.bareMetalOnly();
            cli.writeerror('Parse server hosting is currently only available in the Bare Metal Cloud environment'); 
            process.exit(1); 
        }

        return endpoints.baseUrl(__env) + 'v1/app/deploycloudcode';
    }, 
    login: (process.env.DEVELOP ? BARE_METAL_CLOUD_DEV : BARE_METAL_CLOUD_PROD) + 'v1/account/getapikey',
    status: function(__env) {
        return endpoints.baseUrl(__env) +
            ((__env == BARE_METAL_CLOUD_ALIAS) ? 'v1/app/deploymentstatus' : 'instances/app/deploystatus');
    },
    exec : function(__env){
        return endpoints.baseUrl(__env) + ((__env == BARE_METAL_CLOUD_ALIAS) ? 'v1/app/exec' : 'instances/app/exec');
    }, 
    logs: function (__env) {
        return endpoints.baseUrl(__env) +
            ((__env == BARE_METAL_CLOUD_ALIAS) ? 'v1/app/logs' : 'instances/app/logsaggregate');
    },
    appstatus: function (__env) {
        return endpoints.baseUrl(__env) +
            ((__env == BARE_METAL_CLOUD_ALIAS) ? 'v1/status' : 'instances');
    },
    apprestart: function (__env) {
        return 'v1/app/restart'
    },
    dbrestart: function(__env){
        return 'v1/db/restart'; 
    },
    stop: function (__env) {
        return 'v1/app/stop'
    }
};

module.exports = endpoints;
