// nodechef-cli/remotecommon.js
var request = require('request'), 
    endpoints = require('./endpoints'), 
    cli = require('./cli');

var common = {
    stickyCookie : null,
    getSticky : function(res){
        for (var s in res.headers) {
            if (s == 'set-cookie') {
                var cookies = res.headers[s];
                for (var i = 0; i < cookies.length; i++) {
                    if (cookies[i].indexOf('route=') > -1) {
                        common.stickyCookie = cookies[i]; 
                    }
                }
            }
        }
    }, 
    getRequestHeaders: function () {
        var obj = {
            'User-Agent': global.__NC_CLI,
            'Authorization': 'Bearer ' + global.__NC_API_TOKEN
        };

        if (common.stickyCookie != null) {
            obj['Cookie'] = common.stickyCookie; 
        }

        cli.debug(JSON.stringify(obj)); 

        return obj; 
    },
    GetBridgeRoute: function (app_name, cb) {
        request({
            url: endpoints.appCloudEnv(),
            qs: { cname :  app_name},
            method: 'GET',
            headers: common.getRequestHeaders()
        }, function (err, res, body) {
            if ((res && res.statusCode != 200) || !res) {
                if (!err)
                    err = cli.errDef.remoteNon200(res.statusCode);
            }

            if (err || !body) {
                cli.writeerror(err);
                return;
            }

            cli.debug(body); 
            common.getSticky(res);
            var obj = JSON.parse(body);
            cb(obj.cloud);
        });
    },
    ExpectsOk : function(url, query_str, method){
        request({
            url: url,
            qs: query_str,
            method: method,
            headers: common.getRequestHeaders()
        }, function (err, res, body) {
            if ((res && res.statusCode != 200) || !res) {
                if (!err)
                    err = cli.errDef.remoteNon200(res.statusCode);
            }

            if (err || !body) {
                cli.writeerror(err);
                process.exit(1);
            }

            cli.writeline('Ok');
        });
    },
    ByAppName: function (url, application_name, app_id, cmd, params, method, cb) {
        cli.debug('Request url: ' + url);
        var query_str = {
            cname: application_name
        };

        for (var s in params)
            query_str[s] = params[s];

        cli.debug('Request headers: ' + JSON.stringify(query_str));

        request({
            url: url,
            qs: query_str,
            method: method,
            headers: common.getRequestHeaders()
        }, function (err, res, body) {
            if ((res && res.statusCode != 200) || !res) {
                if (!err)
                    err = cli.errDef.remoteNon200(res.statusCode);
            }

            if (err || !body) {
                cli.writeerror(err);
                return;
            }

            common.getSticky(res);
            cb(body);
        });
    },
    resolveAppId: function (application_name, app_id, cmd, cb) {
        if (application_name == null) {
            if (app_id == null) {
                cli.errors.expectsName(cmd);
                process.exit(1);
            }

            cb(app_id);
        }
        else {
            common.getAppId(application_name, function (_ids) {
                if (_ids.length > 1) {
                    cli.writeerror('Application with name ' + cli.writevariable(application_name) + ' has multiple associated instances');
                    cli.writeerror('Available Instances:');

                    for (var i = 0; i < _ids.length; i++) {
                        cli.writeerror(cli.writeerror(_ids[i]));
                    }

                    return;
                }
                
                cb(_ids[0]); 
            })
        }
    }, 
    getAppId: function (name, cb_s, cb_err) {
        if (!cb_s)
            return;

        var _h = common.getRequestHeaders();

        request.get({
            url: endpoints.appId(),
            qs : {
                name : name
            }, 
            headers: _h
        }, function (err, res, body) {
            if ((res && res.statusCode != 200) || !res) {
                if (!err)
                    err = 'Remote request returned an error response code (' + res.statusCode.toString() + ')';

            }

            if (err) {
                cli.writeerror(err);
                if (cb_err) {
                    cb_err();
                }
            }
            else if (body) {
                common.getSticky(res);
                cli.debug('Remote response: ' + body);
                var obj = JSON.parse(body);
                obj = obj['ids']; 
                cb_s(obj); 
            }
        }); 
    }
}

module.exports = common;
